### PR TITLE
Improve chat (+ use userRepository instead of firestore)

### PR DESCRIPTION
**Pull request by lumapools**
Fixes: #255 and #256 
Two modifications have been made:

1. The refactoring of how the chat works: We were using `FirebaseFirestore` inside `ChatsFragment` and `MessageFragment`, but now all this work is dedicated to `userRepository` (and its extents). This way, the code is cleaner to understand and the tests don't depend on `FirebaseFirestore` anymore (`InMemoryUserRepository` updated to handle messaging for tests).

2. The number of unread messages are shown next to each chat partner's username. When a person is on the list of chats, and an existing message partner messages them, the counter increases in real time. If they open the conversation, the counter resets and the red indicator disappears.

resized-image-Promo

Note: This feature works on two separate phones chatting with each other. Also, the chat partners' list doesn't follow any particular order. This is planned for a later Pull Request.

### CODE REVIEW

File: app/src/main/java/com/example/sharingang/database/cache/CachedUserRepository.kt [hunk 1]
```diff
@@ -73,4 +75,32 @@ class CachedUserRepository @Inject constructor(
...
+        return thenRefresh { store.setupRefresh(userId, with, action) }
+    }
+
+    override suspend fun getNumUnread(userId: String, with: String): Long {
+        return thenRefresh { store.getNumUnread(userId, with) }
+    }
+
+    override suspend fun clearNumUnread(userId: String, with: String) {
+        return thenRefresh { store.clearNumUnread(userId, with) }
+    }
```
Comment (betrisey) [inline app/src/main/java/com/example/sharingang/database/cache/CachedUserRepository.kt:105]: I don't think that the `thenRefresh` are necessary what's changed is not stored in the cache

File: app/src/main/java/com/example/sharingang/database/store/UserStore.kt [hunk 2]
```diff
@@ -58,4 +59,57 @@ interface UserStore {
...
+    suspend fun putMessage(from: String, to: String, message: String): MutableList<Chat>
+
+    /**
+     * Sets up the listener on messages for a particular user with another user
+     *
+     * @param userId the current user id
+     * @param with the target user id
+     * @param action what to do upon getting notified
+     */
+    suspend fun setupRefresh(userId: String, with: String, action: () -> Unit)
```
Comment (betrisey) [inline app/src/main/java/com/example/sharingang/database/store/UserStore.kt:97]: Would be nice to reference the chat in the name of the method.
Maybe `setupChatRefresh` or `setupConversationRefresh`

**Review by betrisey — CHANGES_REQUESTED** [review]
Good job!

File: app/src/main/java/com/example/sharingang/database/store/UserStore.kt [hunk 3]
```diff
@@ -58,4 +59,57 @@ interface UserStore {
...
+
+    /**
+     * Puts a message into database
+     *
+     * @param from the sender
+     * @param to the receiver
+     * @param message the message
+     * @return the new list of messages between the two users
+     */
+    suspend fun putMessage(from: String, to: String, message: String): MutableList<Chat>
```
[the remark this answers is not in the document]
Comment (lumapools) [reply app/src/main/java/com/example/sharingang/database/store/UserStore.kt:88]: Okay! I will try to do that!

File: app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt [hunk 4]
```diff
@@ -69,4 +89,130 @@ class FirestoreUserStore @Inject constructor(private val firestore: FirebaseFire
...
     }
+
+    override suspend fun getChatPartners(userId: String): MutableList<String> {
+        val partnerIds = mutableListOf<String>()
+        val chatPartners = firestore.collection(DatabaseFields.DBFIELD_USERS)
+            .document(userId).collection(DatabaseFields.DBFIELD_MESSAGEPARTNERS).get()
+            .await()
+        val listOfDocuments = chatPartners.documents
+        listOfDocuments.forEach { partnerIds.add(it.id) }
+        return partnerIds
```
[the remark this answers is not in the document]
Comment (lumapools) [reply app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt:100]: Okay, I will try!

File: app/src/main/java/com/example/sharingang/database/cache/CachedUserRepository.kt (hunk 1 above)
Comment (betrisey) [reply app/src/main/java/com/example/sharingang/database/cache/CachedUserRepository.kt:105]: Yes (and check if it still works, I'm not 100% sure)

**Review by betrisey — APPROVED** [review]
Thanks :)

File: app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt [hunk 5]
```diff
@@ -69,4 +89,120 @@ class FirestoreUserStore @Inject constructor(private val firestore: FirebaseFire
...
+    override suspend fun getChatPartners(userId: String): List<String> {
+        val chatPartners = firestore.collection(DatabaseFields.DBFIELD_USERS)
+            .document(userId).collection(DatabaseFields.DBFIELD_MESSAGEPARTNERS).get()
+            .await()
+        return chatPartners.documents.map { it.id }
+    }
+
+    override suspend fun getMessages(userId: String, with: String): List<Chat> {
+        val documents = firestore.collection(DatabaseFields.DBFIELD_USERS)
+            .document(userId).collection(DatabaseFields.DBFIELD_CHATS)
```
Comment (bafavre) [inline app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt:102]: Use getUserDocument here.

File: app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt [hunk 6]
```diff
@@ -69,4 +89,120 @@ class FirestoreUserStore @Inject constructor(private val firestore: FirebaseFire
             .collection(DatabaseFields.DBFIELD_REPORTS).document(reporterId).get().await()
         return docIdRef.exists()
     }
+
+    override suspend fun getChatPartners(userId: String): List<String> {
+        val chatPartners = firestore.collection(DatabaseFields.DBFIELD_USERS)
+            .document(userId).collection(DatabaseFields.DBFIELD_MESSAGEPARTNERS).get()
```
Comment (bafavre) [inline app/src/main/java/com/example/sharingang/database/firestore/FirestoreUserStore.kt:95]: Use getUserDocument here.

**Review by bafavre — APPROVED** [review]
Well done, good job! :)

**Pull request merged**